airbyte-integrations/connectors/source-google-search-console/source_google_search_console/streams.py

```diff
@@ -0,0 +1,303 @@
+
+
+class SearchAnalytics(GoogleSearchConsole, ABC):
+    """
+    API docs: https://developers.google.com/webmaster-tools/search-console-api-original/v3/searchanalytics
+    """
+
+    data_field = "rows"
+    start_row = 0
+    value_field = None
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        refresh_token: str,
+        site_urls: list,
+        start_date: pendulum.datetime,
+        service_account_info: str,
+    ):
+        super().__init__(client_id, client_secret, refresh_token, site_urls, start_date, service_account_info)
+        self.search_types = ["web", "news", "image", "video"]
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        return f"/sites/{self._site_urls[0]}/searchAnalytics/query"
+
+    @property
+    def cursor_field(self) -> Union[str, List[str]]:
+        return "date"
+
+    @property
+    def http_method(self) -> str:
+        return "POST"
+
+    def next_page_token(self, response: requests.Response) -> Optional[dict]:
+        result = {}
+
+        if len(response.json().get(self.data_field, [])) == ROW_LIMIT:
+            self.start_row += ROW_LIMIT
+
+        elif len(self.search_types) > 1:
+            self.search_types.pop(0)
+            self.start_row = 0
+
+        elif len(self._site_urls) > 1:
+            self._site_urls.pop(0)
+            self.search_types = ["web", "news", "image", "video"]
+            self.start_row = 0
+
+        else:
+            return None
+
+        result["starRow"] = self.start_row
+        result["searchType"] = self.search_types[0]
+
+        return result
+
+    def request_headers(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Mapping[str, Any]:
+        headers = {"Content-Type": "application/json"}
+        return headers
+
+    def request_body_json(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Optional[Union[Dict[str, Any], str]]:
+
+        start_date = self._start_date
+        if start_date and stream_state:
+            if stream_state.get(self.cursor_field):
+                start_date = max(
+                    pendulum.parse(stream_state[self.cursor_field]),
+                    pendulum.parse(start_date),
+                ).to_date_string()
+
+        data = {
+            "startDate": start_date,
+            "endDate": stream_state.get("end_date") or pendulum.now().to_date_string(),
+            "dimensions": ["date", self.value_field],
+            "searchType": next_page_token.get("searchType") if isinstance(next_page_token, dict) else "web",
```

Comment:
this would be a lot cleaner with stream slices to avoid this `or` logic. If I change the order of strings in `search_types` array then this logic is incorrect